modules/createServerRenderContext.js

```diff
@@ -1,34 +1,36 @@
-const k = () => {}
-
 const createServerRenderContext = () => {
   let flushed = false
+  let flushedIndex = 0
   let redirect = null
   let matchContexts = [
-    /* { hasMissComponent: bool, matchesByIdentity: [] } */
+    /* { hasMissComponent: bool, hasMatches: bool } */
   ]
 
-  const setRedirect = flushed ? k : (location) => {
+  const setRedirect = (location) => {
     if (!redirect)
       redirect = location
   }
 
-  const registerMatchContext = flushed ? k : (matchesByIdentity) => {
-    return matchContexts.push({
-      hasMissComponent: false,
-      matchesByIdentity
-    }) - 1
+  // on the second pass, index is determined by render order
+  const registerMatchProvider = () => {
+    return flushed ? flushedIndex++ : (
```

Comment:
On the first pass, each `<MatchProvider>` adds an item to the `matchContexts` arrays. This is done through the `registerMatchProvider` function, which returns the index of the added item. That index is stored by the `<MatchProvider>` as `serverRouterIndex`. While the first pass is rendering, it will use the index to manipulate the item in order to indicate if it a) has any matches and b) has any `<Miss>` components.

On the second pass, we want to reference the items in the array that were created during the first pass. For each `<MatchProvider>`, we receive the index for the correct item in the `matchContexts` array via the call to `registerMatchProvider`. This is done using `flushedIndex` variable, simply incrementing the index returned on each call (that's why the render order is important).